scheduling/taskmetadata.go

```diff
@@ -0,0 +1,463 @@
+		if e != nil {
+			return "", fmt.Errorf("failed to unmarshal metadata: %v", err)
+		}
+		if status == Failed && err == nil {
+			return "", fmt.Errorf("error is required for failed status")
+		}
+		metadata.Status = status
+		if err == nil {
+			metadata.Error = ""
+		} else {
+			metadata.Error = err.Error()
+		}
+
+		serializedMetadata, e := metadata.Marshal()
+		if e != nil {
+			return "", fmt.Errorf("failed to marshal metadata: %v", e)
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storer.Update(taskRunMetadataKey.String(), updateStatus)
+	return e
+}
+
+func (m *TaskMetadataManager) SetRunEndTime(runID TaskRunID, taskID TaskID, time time.Time) error {
+	if taskID <= 0 {
+		return fmt.Errorf("invalid run id: %d", taskID)
+	}
+	if time.IsZero() {
+		return fmt.Errorf("invalid run end time: %v", time)
+	}
+
+	metadata, e := m.GetRunByID(taskID, runID)
+	if e != nil {
+		return fmt.Errorf("failed to fetch run: %v", e)
+	}
+
+	updateEndTime := func(runMetadata string) (string, error) {
+		metadata := TaskRunMetadata{}
+		err := metadata.Unmarshal([]byte(runMetadata))
+		if err != nil {
+			return "", fmt.Errorf("failed to unmarshal metadata: %v", err)
+		}
+
+		if metadata.StartTime.After(time) {
+			return "", fmt.Errorf("end time cannot be before start time")
+		}
+
+		metadata.EndTime = time
+		serializedMetadata, e := metadata.Marshal()
+		if e != nil {
+			return "", fmt.Errorf("failed to marshal metadata: %v", e)
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storer.Update(taskRunMetadataKey.String(), updateEndTime)
+	return e
+}
+
+func (m *TaskMetadataManager) AppendRunLog(runID TaskRunID, taskID TaskID, log string) error {
+	if runID <= 0 {
+		return fmt.Errorf("invalid run id: %d", taskID)
+	}
+	if taskID <= 0 {
+		return fmt.Errorf("invalid task id: %d", taskID)
+	}
+	if log == "" {
+		return fmt.Errorf("log cannot be empty")
+	}
+
+	metadata, e := m.GetRunByID(taskID, runID)
+	if e != nil {
+		return fmt.Errorf("failed to fetch run: %v", e)
+	}
+
+	updateLog := func(runMetadata string) (string, error) {
+		metadata := TaskRunMetadata{}
+		err := metadata.Unmarshal([]byte(runMetadata))
+		if err != nil {
+			return "", fmt.Errorf("failed to unmarshal metadata: %v", err)
+		}
+
+		metadata.Logs = append(metadata.Logs, log)
+
+		serializedMetadata, e := metadata.Marshal()
+		if e != nil {
+			return "", fmt.Errorf("failed to marshal metadata: %v", e)
+		}
+
+		return string(serializedMetadata), nil
+	}
+
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: metadata.StartTime}
+	e = m.storer.Update(taskRunMetadataKey.String(), updateLog)
+	return e
+}
+
+// Finds the highest increment in a list of strings formatted like "/tasks/metadata/task_id=0"
```

Comment:
could probably put this in `TaskMetadata` or maybe create a helpers directory for these functions